### PR TITLE
Enable packing for all-gather CP

### PR DESCRIPTION
**Pull request by huytransformer**
# Description

Previously, packed CP training was rejected unless `context_parallel_strategy="ring"`. The all-gather CP already has segment IDs and positions via the packed data, and TPU Flash attention already uses Q sharded over context with K/V gathered for attention.

b/525146921

# Tests

Ran TPU unit test:

```bash
pytest tests/unit/attention_test.py::AttentionTest::test_tpu_flash_attention_packed_all_gather_context_parallel
```

# Checklist

- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by igorts-git — APPROVED** [review]

File: src/maxtext/utils/train_utils.py [hunk 1]
```diff
@@ -247,19 +247,29 @@ def create_train_state_fn():
...
     # Apply reordering wrapper to data iterators if context parallelism is enabled
     with jax.set_mesh(mesh):
       if config.context_parallel_size > 1 and config.context_parallel_load_balance:
 
         # Determine load balancing reorder strategy based on whether packing is enabled
         if config.context_parallel_reorder_strategy == ReorderStrategy.AUTO:
-          reorder_strategy = ReorderStrategy.STRIPED if config.packing else ReorderStrategy.DUAL_CHUNK_SWAP
+          reorder_strategy = (
+              ReorderStrategy.STRIPED
+              if config.packing and config.context_parallel_strategy.lower() == "ring"
```
Comment (richjames0) [inline src/maxtext/utils/train_utils.py]: nit: you could use your context_parallel_strategy defined earlier instead of lowering again

**Review by richjames0 — APPROVED** [review]
lgtm with nit

File: src/maxtext/utils/train_utils.py [hunk 2]
```diff
@@ -247,19 +247,29 @@ def create_train_state_fn():
...
-            "Context parallelism with 'all_gather' strategy cannot be used with sequence packing. "
-            "Please use 'ring' strategy instead."
+            "Context parallelism with sequence packing supports context_parallel_strategy='all_gather' or 'ring'."
         )
+      if (
+          config.hardware in ("gpu", "gpu_multiprocess")
+          and config.attention == "cudnn_flash_te"
+          and not (context_parallel_strategy == "ring" and config.context_parallel_load_balance)
+      ):
+        raise ValueError("Packing is only supported for load balanced ring attention with context parallelism.")
```
Comment (NuojCheng) [inline src/maxtext/utils/train_utils.py]: maybe add ""Packing is only supported for load balanced ring attention with context parallelism for GPU."

File: tests/unit/attention_test.py [hunk 3]
```diff
@@ -865,6 +865,82 @@ def test_tpu_flash_attention_context_parallel(
...
+        decoder_segment_ids,
+        decoder_positions,
+    )
+
+    self.assertTrue(
+        jax.numpy.allclose(
+            jax.device_get(mha_generic_output),
+            jax.device_get(mha_generic_flash_cp_output),
+            rtol=1e-01,
+            atol=1e-01,
```
Comment (NuojCheng) [inline tests/unit/attention_test.py:937]: 1e-1 might be too large, what about 1e-2?

Comment (huytransformer) [reply tests/unit/attention_test.py:937]: It doesnt pass 1e-2 but I checked the error distribution:
```
max: 0.015625
mean: 0.000888824
p99: 0.00390625
p999: 0.0078125
>1e-2: 2.7179718e-05
>1e-1: 0.0
```
there's that tiny long tail that failed the threshold but maybe this is bf16 scale differences (?)

The sister test here also uses 1e-1 tho
`tests/unit/attention_test.py::AttentionTest::test_tpu_flash_attention_context_parallel`

File: src/maxtext/utils/train_utils.py (hunk 1 above)
Comment (huytransformer) [reply src/maxtext/utils/train_utils.py]: agreed!!

File: src/maxtext/utils/train_utils.py (hunk 2 above)
Comment (huytransformer) [reply src/maxtext/utils/train_utils.py]: ok
